File: app/src/main/java/com/bopr/android/smailer/remote/RemoteControlWorker.java
package com.bopr.android.smailer.remote;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.work.Constraints;
import androidx.work.ExistingPeriodicWorkPolicy;
import androidx.work.NetworkType;
import androidx.work.PeriodicWorkRequest;
import androidx.work.WorkManager;
import androidx.work.Worker;
import androidx.work.WorkerParameters;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.concurrent.TimeUnit;

import static com.bopr.android.smailer.Settings.PREF_REMOTE_CONTROL_ENABLED;
import static com.bopr.android.smailer.Settings.settings;
import static com.bopr.android.smailer.remote.RemoteControlService.startRemoteControlService;

/**
 * Periodically checks email out for remote tasks.
 *
 * @author Boris Pronin (<a href="mailto:boprsoft.dev@gmail.com">boprsoft.dev@gmail.com</a>)
 */
public class RemoteControlWorker extends Worker {

    private static final Logger log = LoggerFactory.getLogger("RemoteControlWorker");

    private static final String WORKER_TAG = "smailer-email";

    public RemoteControlWorker(@NonNull Context context, @NonNull WorkerParameters workerParams) {
        super(context, workerParams);
    }

    @NonNull
    @Override
    public Result doWork() {
        log.debug("Working");

        Context context = getApplicationContext();
        if (isFeatureEnabled(context)) {
            startRemoteControlService(context);
        }
        return Result.success();
    }

    private static boolean isFeatureEnabled(@NonNull Context context) {
        return settings(context).getBoolean(PREF_REMOTE_CONTROL_ENABLED, false);
    }

    public static void enableRemoteControlWorker(@NonNull Context context) {
        WorkManager manager = WorkManager.getInstance();

        manager.cancelAllWorkByTag(WORKER_TAG);

        if (isFeatureEnabled(context)) {
            Constraints constraints = new Constraints.Builder()
                    .setRequiredNetworkType(NetworkType.CONNECTED)
                    .build();
            PeriodicWorkRequest request = new PeriodicWorkRequest.Builder(RemoteControlWorker.class,
                    15, TimeUnit.MINUTES) /* interval must be lesser than PeriodicWorkRequest.MIN_PERIODIC_INTERVAL_MILLIS */
                    .addTag(WORKER_TAG)
                    .setConstraints(constraints)
                    .build();
            manager.enqueueUniquePeriodicWork(WORKER_TAG, ExistingPeriodicWorkPolicy.REPLACE, request);

            log.debug("Enabled");
        } else {
            log.debug("Disabled");
        }
    }
}

File: app/src/main/kotlin/com/bopr/android/smailer/remote/RemoteControlWorker.kt
package com.bopr.android.smailer.remote

import android.content.Context
import androidx.work.*
import androidx.work.ExistingPeriodicWorkPolicy.*
import androidx.work.NetworkType.CONNECTED
import androidx.work.PeriodicWorkRequest.Companion.MIN_PERIODIC_INTERVAL_MILLIS
import com.bopr.android.smailer.Settings
import org.slf4j.LoggerFactory
import java.util.concurrent.TimeUnit.MILLISECONDS

/**
 * Periodically checks email out for remote tasks.
 *
 * @author Boris Pronin ([boprsoft.dev@gmail.com](mailto:boprsoft.dev@gmail.com))
 */
internal class RemoteControlWorker(context: Context, workerParams: WorkerParameters) :
    Worker(context, workerParams) {

    override fun doWork(): Result {
        applicationContext.run {
            if (isFeatureEnabled()) {
                RemoteControlProcessor(this).checkMailbox()
            }
        }

        return Result.success()
    }

    internal companion object {

        private val log = LoggerFactory.getLogger("RemoteControlWorker")
        private const val WORK_REMOTE = "com.bopr.android.smailer.remote"

        private fun Context.isFeatureEnabled() = Settings(this).isRemoteControlEnabled

        fun Context.enableRemoteControl() {
            val manager = WorkManager.getInstance(this)
            if (isFeatureEnabled()) {
                log.debug("Enabled")

                val constraints = Constraints.Builder()
                    .setRequiredNetworkType(CONNECTED)
                    .build()
                val request = PeriodicWorkRequest.Builder(
                    RemoteControlWorker::class.java,
                    MIN_PERIODIC_INTERVAL_MILLIS, MILLISECONDS
                )
                    .setConstraints(constraints)
                    .build()
                manager.enqueueUniquePeriodicWork(WORK_REMOTE, UPDATE, request)
            } else {
                log.debug("Disabled")

                manager.cancelUniqueWork(WORK_REMOTE)
            }
        }
    }
}